File: articles/life-death-problem-of-biological-definition/python/host_virus_dynamics.py
"""
Host-virus dynamics workflow.

Run:
    python python/host_virus_dynamics.py
"""

from __future__ import annotations

from pathlib import Path

import numpy as np
import pandas as pd

from definition_core import simulate_host_virus


ARTICLE_DIR = Path(__file__).resolve().parents[1]
VIRUS_PATH = ARTICLE_DIR / "data" / "host_virus_scenarios.csv"


def main() -> None:
    """Simulate host-virus scenarios."""

    scenarios = pd.read_csv(VIRUS_PATH)

    rows = []

    for _, scenario in scenarios.iterrows():
        time = np.arange(0, scenario["time_end"] + scenario["dt"], scenario["dt"])

        sim = simulate_host_virus(
            time=time,
            target_initial=scenario["target_initial"],
            infected_initial=scenario["infected_initial"],
            virus_initial=scenario["virus_initial"],
            beta=scenario["beta"],
            delta=scenario["delta"],
            production=scenario["production"],
            clearance=scenario["clearance"],
        )

        rows.append(
            {
                "scenario": scenario["scenario"],
                "final_target_cells": sim["target_cells"][-1],
                "final_infected_cells": sim["infected_cells"][-1],
                "final_free_virus": sim["free_virus"][-1],
                "peak_infected_cells": sim["infected_cells"].max(),
                "peak_free_virus": sim["free_virus"].max(),
                "time_to_peak_virus": time[sim["free_virus"].argmax()],
            }
        )

    print(pd.DataFrame(rows).round(5).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/life-death-problem-of-biological-definition/r/host_virus_dynamics.R
# Simple host-virus dynamics in R.

scenario_path <- file.path("data", "host_virus_scenarios.csv")

if (!file.exists(scenario_path)) {
  scenario_path <- file.path("..", "data", "host_virus_scenarios.csv")
}

simulate_host_virus <- function(target_initial, infected_initial, virus_initial, beta, delta, production, clearance, time_end, dt) {
  time <- seq(0, time_end, by = dt)

  target <- numeric(length(time))
  infected <- numeric(length(time))
  virus <- numeric(length(time))

  target[1] <- target_initial
  infected[1] <- infected_initial
  virus[1] <- virus_initial

  for (i in 2:length(time)) {
    d_target <- -beta * target[i - 1] * virus[i - 1]
    d_infected <- beta * target[i - 1] * virus[i - 1] - delta * infected[i - 1]
    d_virus <- production * infected[i - 1] - clearance * virus[i - 1]

    target[i] <- max(target[i - 1] + d_target * dt, 0)
    infected[i] <- max(infected[i - 1] + d_infected * dt, 0)
    virus[i] <- max(virus[i - 1] + d_virus * dt, 0)
  }

  data.frame(
    time = time,
    target_cells = target,
    infected_cells = infected,
    free_virus = virus
  )
}

scenarios <- read.csv(scenario_path)

rows <- lapply(seq_len(nrow(scenarios)), function(i) {
  s <- scenarios[i, ]

  sim <- simulate_host_virus(
    s$target_initial,
    s$infected_initial,
    s$virus_initial,
    s$beta,
    s$delta,
    s$production,
    s$clearance,
    s$time_end,
    s$dt
  )

  data.frame(
    scenario = s$scenario,
    final_target_cells = tail(sim$target_cells, 1),
    final_infected_cells = tail(sim$infected_cells, 1),
    final_free_virus = tail(sim$free_virus, 1),
    peak_infected_cells = max(sim$infected_cells),
    peak_free_virus = max(sim$free_virus)
  )
})

summary_df <- do.call(rbind, rows)

print(round(summary_df, 5))
